src/views/QuantumTP6/index.jsx

```diff
@@ -31,36 +32,39 @@ class TP6 extends React.Component {
     const { classes } = this.props;
     const { bits } = this.state;
 
-    return (
-      <div className={classes.body}>
-        <DashboardPage
-          key={bits}
-          title="TP6"
-          subtitle={`Page load on ${bits} bits`}>
-          <Grid container spacing={24}>
-            {frum(TP6_PAGES)
-              .where({ bits })
-              .groupBy('title')
-              .map((series, title) => (
-                <Grid
-                  item
-                  xs={6}
-                  key={`page_${title}_${bits}`}
-                  className={classes.chart}>
-                  <PerfherderGraphContainer
-                    title={title}
-                    series={frum(series)
-                      .sortBy(['browser'])
-                      .reverse()
-                      .map(s => ({ label: s.label, seriesConfig: s }))
-                      .toArray()}
-                  />
-                </Grid>
-              ))}
-          </Grid>
-        </DashboardPage>
-      </div>
-    );
+    if (bits === '32' || bits === '64')
```

Comment:
I am wondering out loud:

It seems messy to pollute the component with these checks.  Not just here, but it is happening in other parts of this PR too,  Can we let the Component throw an error, and let an HOC show the `InvalidUrlMessage`?  We may need to distinguish between user error (like in this case with bad URL)  and network error (which happens often to me).

We must still do a verification, but it need only be one line:

    if (bits === '32' || bits === '64') throw new Error("bad bits!");

I am currently attempting an HOC here https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/313/files#diff-21b4dce6876c43d2d41fa75031291026R11

Reply:
@klahnakoski I think I got what you're trying to say. But for my clarification, you mean that we should make a HOC and use that `<InvalidUrlMessage />` component there based on the message returned by wrapped component. 

>We may need to distinguish between user error (like in this case with bad URL) and network error (which happens often to me).

I didn't get it much. May be we need to use HOC only where there is user error ? 

Looking into your HOC, looks like time has come to dig into HOC now :-)

Reply:
> > We may need to distinguish between user error (like in this case with bad URL) and network error (which happens often to me).
> 
> I didn't get it much. May be we need to use HOC only where there is user error ?

@aimenbatool Creating a HOC is one possible solution in general - distinguishing between types of errors isn't specific to using a HOC. I suggest you read up on the advantages and potential pitfalls of using HOCs (not just understanding how they work) because it's important to understand the trade-offs of using one approach over another. :)

There are a few approaches you can take to handle user errors (http 400) - whether the user entered an invalid value or query parameter:
1) verify a value/parameter is valid _before_ making an API request and show a useful message to the user (for example, you can check for valid date formats or character length or whether a value is a number). With react router you can access parameters with the `location` prop.
2) You won't be able to verify every query parameter before it is used to make a request, and so the API may return a http 400 - bad request or http 404 - not found. Depending on how the API is configured, it should return a (hopefully) useful json message that can be shown to the user. Here's [an example ](https://treeherder.mozilla.org/api/performance/summary/ ) with `performance/summary` API.

The Fetch API has a few limitations regarding error handling, in that it will only reject a Promise for network related errors (5**) - it won't reject 4** errors. One way around this is to create a wrapper around `fetch` and check for the response.ok. If there is a 4** error that is likely returning a json response for the error, you can process that accordingly. For 5** errors, whether a json or text/html response is returned might be based on how the server is configured. You can find several articles on this topic (http status codes and how/what to show to the user) on the internet. It's worth reading up on this topic generally since there are different ways to approach it.

Reply:
@sarah-clements  please, feel free to contradict me; At the very least, there is information in our disagreement that I can benefit from. 

You bring up this is a routing issue: Maybe the error handling belongs in `<BrowserRouter>` as some more specific URL rules.  Maybe the error handling belongs in `withNavigation()`: which is responsible for URL parameters. Maybe it belongs in an HOC: that attempts to do its best to provide a humane feedback when failure happens.  Maybe error handling belongs on all three, or belongs in none.  

Your points about HOCs getting tricky may be valid.  HOCs should not be changing behaviour of the child components so I do not think this is true. However, I have not used HOCs enough to say if that is an unrealistic idealism, so you correct to point out there may be dragons there.

On the very specific subject of long HOC lines, like your example:

    withUserError(withLoading(withRouter(withStyles(styles)(Output))))

I do not think this long line is a problem; it think it is just aside effect of functional notation being a programmer's default, and not much consideration given to long chains of HOCs.  I would expect we eventually move to chaining the HOCs for readability.

```js
with(Output)
  .withStyles(styles)
  .withNavigation(config)
  .withRouter()
  .withUserError() 
;
```


At the very least, I would like to *attempt* removing this pattern from our components:

```js
error ? <ErrorComponent> : <GoodComponent>
```

That pattern has various forms, but it seems to pollute the code because it is everywhere.  I can be wrong about this because it is frontend code.  For backend code, when I see an `if` statement, it often indicates a bad data model. Many times these models are out of your control, so `if` statements are required; otherwise, we should be inspecting **why** the data is irregular and if we can fix it; if we can not fix it, because it is fundamental to the domain, then we should encapsulate the irregularity in a more regular data structure. 

For example, I am looking at the perfherder code so we can pull more data with less requests:  Here is each if statment and how to deal with it.

---
https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L27

This is filtering results, it is better to show it as such:

    toPairs(signatures).where({suite})....

---

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L43

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L45

These two are also filtering, but the hard way: matching hashes, then filtering the options found. The data structure should have joined the two tables on the hash, and then added to the filter above:

    toPairs(signatures).where({suite, option})....

---

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L53

I consider this a legitimate check because it is an `assert`: We should be able to remove this code, but we do not for safety.

---

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L87

Assume `percentileThreshold <= 100` and remove this if statement.

---

https://github.com/mozilla-frontend-infra/firefox-health-dashboard/blob/951b50ff431fbc3f2fe181437931b0940c7941ea/src/utils/perfherder/subbenchmarks.jsx#L111

The data coming out of Perfherder has an irregularity: The aggregate of the subtests is treated like a subtest. So we have if statements that checks for this, and you can see them littered throughout the code.  Even so, this `if` should not exist: any caller must deal with the aggregate separate from the subtest, and will have a symbiotic `if` statement on it's end; making *two* if statements to handle the same data irregularity.  I am guessing this `if` exists so one perfherder request is made instead of two. In this case, I suggest changing the code so the caller decides the signatures needed, and let this function simply retrieve them blindly.

---

Reply:
BTW, I do not intend to block this PR. If it works, then it is good enough for me!  The good thing about working code, beyond the fact it *works*; it represents use cases for later refactoring efforts.  We can attempt other error-handling strategies later.

Reply:
> You bring up this is a routing issue: Maybe the error handling belongs in `<BrowserRouter>` as some more specific URL rules. Maybe the error handling belongs in `withNavigation()`: which is responsible for URL parameters. Maybe it belongs in an HOC: that attempts to do its best to provide a humane feedback when failure happens. Maybe error handling belongs on all three, or belongs in none.

I think this would be a good use case for having an alternate view in `<BrowserRouter>` that shows 'Not found - did you mean x or x?' (or some other implementation). 

> I do think your points about HOCs getting tricky may be valid. HOCs should not be changing behaviour of the child components so I do not think this is true. However, I have not used HOCs enough to say if that is an unrealistic idealism, so you correct to point out there may be dragons there.

I'm not opposed to them generally - I have used it myself where there was stateful logic that could be shared with two different views, but I didn't want to hardwire the layout because there were some differences and I wanted to maintain flexibility.  The apprehension was regarding using multiple HOC for different aspects of error handling that could have the potential for prop name collisions and might be more efficient to combine.

> That pattern has various forms, but it seems to pollute the code because it is everywhere. I can be wrong about this because it is frontend code. For backend code, when I see an `if` statement, it often indicates a bad data model. Many times these models are out of your control, so `if` statements are required; otherwise, we should be inspecting **why** the data is irregular and if we can fix it; if we can not fix it, because it is fundamental to the domain, then we should encapsulate the irregularity in a more regular data structure.
> 
Yes, since this particular logic is reproduced everywhere so it does make sense to encapsulate it if it can be done simply. I've had these thoughts myself regarding error handling in Treeherder. I see a challenge in how much of the different error handling to unite and which to keep separate (with ErrorBoundary components, query param validation, API request failures, etc!)

I appreciate having these discussions and I think it's also useful for @aimenbatool to be able to see the 'why' behind a decision. I'm a fan of starting simply and refactoring as needed, which was why I was trying to encourage that direction for her. I think it'd be a great learning opportunity to land this working code as you've suggested above, and for her to work with you on different error handling strategies :)

Reply:
Great discussion @sarah-clements and @klahnakoski! 
Although it took me time to absorb and try to understand, it is good to see how teams discuss things and make decisions. Somethings are still alien and overwhelming to me in the discussion as a newbie. 
I would like to pick the suggestion 'land this code and refactor step by step'. It will help me not to overwhelm and get confused with a lot of things coming at the same time. :-)  I would love to work on error handling strategies but  first I need to get my self more comfortable playing with HOCs on simple things.